xc/xenops_server_xen.ml

```diff
@@ -282,12 +284,12 @@ let create_vbd_frontend ~xc ~xs task frontend_domid vdi =
       (* There's no need to use a PV disk if we're in the same domain *)
       let _xendisks, blockdevs, files, nbds = Storage_interface.implementations_of_backend vdi.attach_info in
       match files, blockdevs, nbds with
-      | {path}::_, _, _ | _, {path}::_, _ -> Name path
       | _, _, nbd::_ -> Nbd nbd
+      | {path}::_, _, _ | _, {path}::_, _ -> Name path
```

Comment:
Opening a file (note: not a blockdev) directly would be more efficient than going through NBD, any reason why NBD is preferred here? IMHO it just adds an extra layer and potentially more memory usage because in some cases we need to spawn an NBD client. With a file we can do O_DIRECT I/O without any extra memory usage.

Perhaps this should be split:
```ocaml
| {path}::_, _, _ -> Name path
| _, _, nbd :: _ -> Nbd nbd
| _, _, {path}:_ -> Name path
```

Although perhaps it would be easier if SMAPIv1 did not offer a blockdev implementation at all. Then we can leave everything as is (in theory a blockdev that points to lets say /dev/sdb1 would be more efficient than going through NBD to talk to the same blockdev, it is just tapdisk's blockdev that we don't like ....)

In fact I'm not even sure how SMAPIv1 would return you an NBD, because XAPI has it hardcoded to return a XenDisk + BlockDevice:
```ocaml
                 implementations = [
                   XenDisk {
                     params = attach_info_v1.Smint.params;
                     extra = attach_info_v1.Smint.xenstore_data;
                     backend_type = "vbd3"
                   };
                   (* Currently we always get a BlockDevice from SMAPIv1, never a File, not even for ISOs *)
                   BlockDevice {
                     path = attach_info_v1.Smint.params
                   }
                 ]
```

Were you able to get a SMAPIv1 SR to talk NBD?